File: src/checks/audio/tst_trackactivity.cpp
#include "checks/audio/tst_trackactivity.h"
#include "checks/fwd.hpp"

#include <QtTest>

#include <cmath>

#include "ui/activity/trackactivity.h"

namespace checks {
namespace {

constexpr float kTolerance = 0.0001f;
constexpr int kIsolatedTrack = 3;

bool nearlyEqual(float actual, float expected)
{
    return std::abs(actual - expected) <= kTolerance;
}

bool dark(TrackActivityIntensity intensity)
{
    return intensity.left == 0.0f && intensity.right == 0.0f;
}

bool everyTrackDark(const TrackActivity &activity)
{
    for (int track = 0; track < int(kMaxTracks); ++track) {
        if (!dark(activity.intensity(track)))
            return false;
    }
    return true;
}

} // namespace

void TrackActivityTest::freshActivityIsDark()
{
    TrackActivity activity;
    for (int track = 0; track < int(kMaxTracks); ++track)
        QVERIFY2(dark(activity.intensity(track)), "new activity must be dark");
    QVERIFY2(dark(activity.intensity(-1)), "invalid negative track must be dark");
    QVERIFY2(dark(activity.intensity(int(kMaxTracks))), "invalid high track must be dark");
}

void TrackActivityTest::attackRetainsStereoTargets_data()
{
    QTest::addColumn<uint8_t>("peakLeft");
    QTest::addColumn<uint8_t>("peakRight");

    QTest::newRow("asymmetric") << uint8_t{128} << uint8_t{32};
    QTest::newRow("left-only") << uint8_t{255} << uint8_t{0};
    QTest::newRow("right-only") << uint8_t{0} << uint8_t{255};
}

void TrackActivityTest::attackRetainsStereoTargets()
{
    QFETCH(uint8_t, peakLeft);
    QFETCH(uint8_t, peakRight);

    TrackActivity activity;
    TrackActivityLevels levels{};
    levels[kIsolatedTrack] = {peakLeft, peakRight};
    QVERIFY2(activity.advance(levels, 0.017f, true), "playing activity must require another tick");

    const auto first = activity.intensity(kIsolatedTrack);
    // A side with a peak attacks toward it without snapping; a silent side
    // stays exactly at zero (hard-panned peaks never cross-bleed).
    const auto approaches = [](float value, uint8_t peak) {
        return peak > 0 ? (value > 0.0f && value < peak / 255.0f) : value == 0.0f;
    };
    QVERIFY2(approaches(first.left, peakLeft),
             "left attack must approach the peak without snapping");
    QVERIFY2(approaches(first.right, peakRight),
             "right attack must approach the peak without snapping");
    QVERIFY2(peakLeft > peakRight ? first.left > first.right : first.right > first.left,
             "stereo sides must retain independent targets with no channel bleed");
    for (int track = 0; track < int(kMaxTracks); ++track) {
        if (track != kIsolatedTrack)
            QVERIFY2(dark(activity.intensity(track)), "a peak must not light another track");
    }
}

void TrackActivityTest::releaseIsGradualAndMonotonic()
{
    TrackActivity activity;
    TrackActivityLevels levels{};
    levels[kIsolatedTrack] = {128, 32};
    activity.advance(levels, 0.017f, true);

    levels.fill({});
    auto previous = activity.intensity(kIsolatedTrack);
    activity.advance(levels, 0.125f, true);
    auto current = activity.intensity(kIsolatedTrack);
    QVERIFY2(current.left > previous.left * 0.5f && current.right > previous.right * 0.5f,
             "ordinary release must remain visibly gradual");
    previous = current;
    for (auto step = 0; step < 23; ++step) {
        activity.advance(levels, 0.125f, true);
        current = activity.intensity(kIsolatedTrack);
        QVERIFY2(current.left <= previous.left && current.right <= previous.right,
                 "ordinary release must decay monotonically on both sides");
        previous = current;
    }
    QVERIFY2(dark(activity.intensity(kIsolatedTrack)), "released activity must settle at zero");
}

void TrackActivityTest::imperceptibleTailSnapsToZero()
{
    TrackActivity activity;
    TrackActivityLevels levels{};
    levels[kIsolatedTrack] = {255, 255};
    activity.advance(levels, 0.015f, true);
    levels.fill({});
    activity.advance(levels, 2.0f, true);
    QVERIFY2(dark(activity.intensity(kIsolatedTrack)),
             "the visible floor must remove an imperceptible release tail");
}

void TrackActivityTest::retriggerRetainsInertia()
{
    constexpr int kRetriggerTrack = 7;
    TrackActivity activity;
    TrackActivityLevels levels{};

    levels[kRetriggerTrack] = {48, 0};
    activity.advance(levels, 0.017f, true);
    levels.fill({});
    activity.advance(levels, 1.0f, true);
    const auto decayed = activity.intensity(kRetriggerTrack);

    levels[kRetriggerTrack] = {224, 64};
    activity.advance(levels, 0.017f, true);
    const auto retriggered = activity.intensity(kRetriggerTrack);
    QVERIFY2(retriggered.left > decayed.left && retriggered.right > decayed.right,
             "a new peak must retrigger each side independently");
    QVERIFY2(retriggered.left < 224.0f / 255.0f && retriggered.right < 64.0f / 255.0f,
             "a retrigger must retain thermal attack inertia");
}

void TrackActivityTest::resetClearsEveryLight()
{
    TrackActivity activity;
    TrackActivityLevels levels{};
    levels[kIsolatedTrack] = {200, 120};
    activity.advance(levels, 0.017f, true);
    activity.reset();
    QVERIFY2(everyTrackDark(activity), "reset must clear every light");
}

void TrackActivityTest::pausedFillSettlesEveryTrack()
{
    TrackActivity activity;
    TrackActivityLevels levels{};

    levels[kIsolatedTrack] = {255, 64};
    activity.advance(levels, 0.015f, true);
    const auto beforePause = activity.intensity(kIsolatedTrack);
    levels.fill({});
    QVERIFY2(activity.advance(levels, 0.015f, false), "paused fill must require another tick");
    const auto filling = activity.intensity(kIsolatedTrack);
    QVERIFY2(filling.left > beforePause.left && filling.right > beforePause.right,
             "paused fill must gently approach full brightness");
    QVERIFY2(!activity.advance(levels, 10.0f, false), "settled paused fill must stop ticks");
    for (int track = 0; track < int(kMaxTracks); ++track) {
        const auto settled = activity.intensity(track);
        QVERIFY2(settled.left == 1.0f && settled.right == 1.0f,
                 "paused fill must settle every side at full brightness");
    }
    QVERIFY2(!activity.advance(levels, 0.0f, false), "settled pause must remain settled");
}

void TrackActivityTest::resumeUsesFastWindowThenOrdinaryRelease()
{
    TrackActivity activity;
    TrackActivityLevels levels{};

    activity.resetPaused();
    QVERIFY2(activity.advance(levels, 0.075f, true), "resuming activity must require another tick");
    const auto resumed = activity.intensity(kIsolatedTrack);
    QVERIFY2(nearlyEqual(resumed.left, std::exp(-0.075f / 0.015f)),
             "resuming descent must use the 75 ms fast window");
    activity.advance(levels, 0.015f, true);
    const auto ordinaryRelease = activity.intensity(kIsolatedTrack);
    QVERIFY2(nearlyEqual(ordinaryRelease.left, resumed.left * std::exp(-0.015f / 0.250f)),
             "release must return to 250 ms after the resume window");
    QVERIFY2(ordinaryRelease.left > resumed.left * std::exp(-0.015f / 0.015f),
             "ordinary release must not retain the fast resume descent");
}

void TrackActivityTest::rapidPauseRearmsFastDescent()
{
    TrackActivity activity;
    TrackActivityLevels levels{};

    activity.resetPaused();
    activity.advance(levels, 0.050f, true);
    activity.advance(levels, 0.001f, false);
    const auto beforeRapidResume = activity.intensity(kIsolatedTrack);
    activity.advance(levels, 0.015f, true);
    const auto rapidResume = activity.intensity(kIsolatedTrack);
    QVERIFY2(rapidResume.left < beforeRapidResume.left * std::exp(-0.015f / 0.250f),
             "a rapid pause/resume must restart fast descent");
}

void TrackActivityTest::clampedElapsedNeedsTicksWithoutMovement()
{
    TrackActivity activity;
    TrackActivityLevels levels{};
    levels[kIsolatedTrack] = {255, 255};
    QVERIFY2(activity.advance(levels, -1.0f, true),
             "playing must require ticks with negative elapsed");
    QVERIFY2(dark(activity.intensity(kIsolatedTrack)), "negative elapsed must not move activity");
    activity.advance(levels, 0.0f, true);
    QVERIFY2(dark(activity.intensity(kIsolatedTrack)), "zero elapsed must not move activity");
}

} // namespace checks

int runTrackActivityCheck(const QStringList &qtArguments)
{
    checks::TrackActivityTest test;
    QStringList arguments{QStringLiteral("trackactivitycheck")};
    arguments.append(qtArguments);
    return QTest::qExec(&test, arguments);
}

File: src/checks/audio/tst_trackactivity.swift
import Foundation
@testable import PorydawApp

func runTrackActivityChecks(_ report: CheckReport) {
    let zero = Array(repeating: AudioActivityLevel(), count: 16)
    func dark(_ value: TrackActivity.Intensity) -> Bool { value.left == 0 && value.right == 0 }
    var activity = TrackActivity()
    let fresh = "TrackActivityTest::freshActivityIsDark"
    for track in 0..<16 {
        report.expect(dark(activity.intensity(track: track)), cppID: fresh, message: "new track is dark")
    }
    report.expect(dark(activity.intensity(track: -1)), cppID: fresh, message: "negative track is dark")
    report.expect(dark(activity.intensity(track: 16)), cppID: fresh, message: "high track is dark")

    for (left, right): (UInt8, UInt8) in [(128, 32), (255, 0), (0, 255)] {
        let id = "TrackActivityTest::attackRetainsStereoTargets"
        activity = TrackActivity()
        var levels = zero
        levels[3] = AudioActivityLevel(left: left, right: right)
        report.expect(activity.advance(levels, elapsedSeconds: 0.017, playing: true),
                      cppID: id, message: "playing requires another tick")
        let first = activity.intensity(track: 3)
        func approaches(_ value: Float, _ peak: UInt8) -> Bool {
            peak > 0 ? value > 0 && value < Float(peak) / 255 : value == 0
        }
        report.expect(approaches(first.left, left), cppID: id, message: "left attacks without snapping")
        report.expect(approaches(first.right, right), cppID: id, message: "right attacks without snapping")
        report.expect(left > right ? first.left > first.right : first.right > first.left,
                      cppID: id, message: "stereo sides retain independent targets")
        for track in 0..<16 where track != 3 {
            report.expect(dark(activity.intensity(track: track)), cppID: id, message: "other tracks remain dark")
        }
    }
    let release = "TrackActivityTest::releaseIsGradualAndMonotonic"
    activity = TrackActivity()
    var levels = zero
    levels[3] = AudioActivityLevel(left: 128, right: 32)
    activity.advance(levels, elapsedSeconds: 0.017, playing: true)
    var previous = activity.intensity(track: 3)
    activity.advance(zero, elapsedSeconds: 0.125, playing: true)
    var current = activity.intensity(track: 3)
    report.expect(current.left > previous.left * 0.5 && current.right > previous.right * 0.5,
                  cppID: release, message: "release remains visibly gradual")
    previous = current
    for _ in 0..<23 {
        activity.advance(zero, elapsedSeconds: 0.125, playing: true)
        current = activity.intensity(track: 3)
        report.expect(current.left <= previous.left && current.right <= previous.right,
                      cppID: release, message: "both sides decay monotonically")
        previous = current
    }
    report.expect(dark(activity.intensity(track: 3)), cppID: release, message: "release settles at zero")

    activity = TrackActivity()
    levels = zero
    levels[3] = AudioActivityLevel(left: 255, right: 255)
    activity.advance(levels, elapsedSeconds: 0.015, playing: true)
    activity.advance(zero, elapsedSeconds: 2, playing: true)
    report.expect(dark(activity.intensity(track: 3)), cppID: "TrackActivityTest::imperceptibleTailSnapsToZero",
                  message: "visible floor removes imperceptible tail")

    activity = TrackActivity()
    levels = zero
    levels[7] = AudioActivityLevel(left: 48)
    activity.advance(levels, elapsedSeconds: 0.017, playing: true)
    activity.advance(zero, elapsedSeconds: 1, playing: true)
    let decayed = activity.intensity(track: 7)
    levels[7] = AudioActivityLevel(left: 224, right: 64)
    activity.advance(levels, elapsedSeconds: 0.017, playing: true)
    let retriggered = activity.intensity(track: 7)
    let retrigger = "TrackActivityTest::retriggerRetainsInertia"
    report.expect(retriggered.left > decayed.left && retriggered.right > decayed.right,
                  cppID: retrigger, message: "new peak retriggers independently")
    report.expect(retriggered.left < 224 / Float(255) && retriggered.right < 64 / Float(255),
                  cppID: retrigger, message: "retrigger retains attack inertia")

    activity = TrackActivity()
    levels = zero
    levels[3] = AudioActivityLevel(left: 200, right: 120)
    activity.advance(levels, elapsedSeconds: 0.017, playing: true)
    activity.reset()
    report.expect((0..<16).allSatisfy { dark(activity.intensity(track: $0)) },
                  cppID: "TrackActivityTest::resetClearsEveryLight", message: "reset clears every light")

    activity = TrackActivity()
    levels[3] = AudioActivityLevel(left: 255, right: 64)
    activity.advance(levels, elapsedSeconds: 0.015, playing: true)
    let beforePause = activity.intensity(track: 3)
    let paused = "TrackActivityTest::pausedFillSettlesEveryTrack"
    report.expect(activity.advance(zero, elapsedSeconds: 0.015, playing: false), cppID: paused,
                  message: "paused fill requires another tick")
    let filling = activity.intensity(track: 3)
    report.expect(filling.left > beforePause.left && filling.right > beforePause.right,
                  cppID: paused, message: "paused fill gently approaches full brightness")
    report.expect(!activity.advance(zero, elapsedSeconds: 10, playing: false), cppID: paused,
                  message: "settled paused fill stops ticks")
    for track in 0..<16 {
        let settled = activity.intensity(track: track)
        report.expect(settled.left == 1 && settled.right == 1, cppID: paused,
                      message: "all sides settle at full brightness")
    }
    report.expect(!activity.advance(zero, elapsedSeconds: 0, playing: false), cppID: paused,
                  message: "settled pause remains settled")

    activity = TrackActivity()
    activity.resetPaused()
    let resume = "TrackActivityTest::resumeUsesFastWindowThenOrdinaryRelease"
    report.expect(activity.advance(zero, elapsedSeconds: 0.075, playing: true), cppID: resume,
                  message: "resume requires ticks")
    let resumed = activity.intensity(track: 3)
    report.expect(abs(resumed.left - exp(-0.075 / Float(0.015))) <= 0.0001, cppID: resume,
                  message: "resume descends through 75ms fast window")
    activity.advance(zero, elapsedSeconds: 0.015, playing: true)
    let ordinary = activity.intensity(track: 3)
    report.expect(abs(ordinary.left - resumed.left * exp(-0.015 / Float(0.250))) <= 0.0001,
                  cppID: resume, message: "release returns to 250ms")
    report.expect(ordinary.left > resumed.left * exp(-0.015 / Float(0.015)), cppID: resume,
                  message: "ordinary release does not retain fast descent")

    activity = TrackActivity()
    activity.resetPaused()
    activity.advance(zero, elapsedSeconds: 0.050, playing: true)
    activity.advance(zero, elapsedSeconds: 0.001, playing: false)
    let beforeRapidResume = activity.intensity(track: 3)
    activity.advance(zero, elapsedSeconds: 0.015, playing: true)
    report.expect(activity.intensity(track: 3).left < beforeRapidResume.left * exp(-0.015 / Float(0.250)),
                  cppID: "TrackActivityTest::rapidPauseRearmsFastDescent", message: "rapid pause restarts descent")

    activity = TrackActivity()
    levels = zero
    levels[3] = AudioActivityLevel(left: 255, right: 255)
    let elapsed = "TrackActivityTest::clampedElapsedNeedsTicksWithoutMovement"
    report.expect(activity.advance(levels, elapsedSeconds: -1, playing: true), cppID: elapsed,
                  message: "negative elapsed still requires ticks")
    report.expect(dark(activity.intensity(track: 3)), cppID: elapsed, message: "negative elapsed does not move")
    activity.advance(levels, elapsedSeconds: 0, playing: true)
    report.expect(dark(activity.intensity(track: 3)), cppID: elapsed, message: "zero elapsed does not move")
}
